### ISSUE TITLE
alfresco/upload/_UploadsDisplayMixin can't be used as a mixin / declare can't build consistent linearization

### ISSUE THREAD

**Issue by AFaust** [Alfresco/Aikau#912]
I tried to build a custom upload progress display (the default one is quite ugly / simplistic) by using alfresco/renderers/Progress as the base and mixin in alfresco/upload/_UploadsDisplayMixin, configuring the widgetsForProgressDialog property on the alfresco/services/UploadService to use the custom progress widget in the dialog. This fails since the "super classes" of alfresco/upload/_UploadsDisplayMixin cause Dojo declare to fail in its super class linearization phase.

Custom widget module (prodyna/upload/SimpleUploadProgress):
```javascript
define([ 'dojo/_base/declare', 'alfresco/renderers/Progress', 'alfresco/upload/_UploadsDisplayMixin'], function(declare, Progress, UploadsDisplayMixin )
{
    return declare([ Progress, _UploadsDisplayMixin ], {/** ... implementation ... */ });
});
```

(Simplified) page JSON configuration:

```javascript
model.jsonModel = {services : [{
        name : 'alfresco/services/UploadService',
        config : {
            widgetsForProgressDialog : [ {
                name : 'prodyna/upload/SimpleUploadProgress',
                assignTo : 'uploadDisplayWidget',
                config : {}
            } ]
        }
    }]};
```

During service initialization, the alfresco_core_CoreWidgetProcessing__processWidgets operation fails when constructing the custom widget with the exception message "declare: can't build consistent linearization"

Since the mixin is "only" supposed to define an interface (JSDoc of the module), it should not include dijit/_WidgetBase nor dijit/_TemplatedMixin as super classes.

This affects Aikau 1.0.51 on Alfresco 5.0.d

**Comment by MartinDoyleUK** [comment]
Good spot. We should definitely let it be an interface if that's what we're calling it. One thing I would suggest in the meantime (we'll raise a bug to fix that) is that you try our new uploader? If you swap out UploadService for FileUploadService, then you'll see it.

**Comment by MartinDoyleUK** [comment]
Have added https://issues.alfresco.com/jira/browse/AKU-889 to address the issue

**Comment by AFaust** [comment]
Just looked at alfresco/upload/UploadMonitor source in 1.0.51 and looks like you "lucked" out on encountering this issue yourself by not including the "usual" base classes. Is there a specific reason why the new "slick" widget isn't used as the default in the UploadService as well? I don't understand why you'd want two UploadService implementations ("old" service isn't yet marked as deprecated if it was the intention to supercede it).

**Comment by MartinDoyleUK** [comment]
It comes from the desire to not break anyone's old implementations. There are obviously benefits to seamlessly improving people's experience with existing Aikau implementations, however they are ultimately just outweighed by ensuring we don't disrupt existing experiences too much.
The intention was really to make life easy for us by having the _UploadsDisplayMixin as a superclass as well as an interface, so the widget base classes were put into it, but you're right in that an interface really should be an interface.

**Comment by draperd** [comment]
Some additional background on this for understanding...

The original UploadService was originally added a long time ago before Aikau had its name and became a separate project. It was part of the original "skunkworks" effort to demonstrate that it would be possible to achieve feature parity with the existing YUI2 Document Library. Rather than simply copying the existing display, I attempted to make it more extendable as a common 3rd party requirement we'd hear was the use case for customizing the upload display (e.g. to maybe perform actions or set metadata on the nodes created after upload had completed).

This original UploadService had no design, and was not required by any part of the product - it also had no tests - and certainly not tests for the customization options.

Since then, the Document Library capabilities have been utilised in our Salesforce Integration solution, and this includes the use of the Document Library - this has resulted in several bugs being fixed and tests added, but **not** for customization of the display.

Following that we have had input resulting from UX reviews and designs for a new uploader - however, this work is ongoing and although almost complete, it took several sprints to implement and we didn't want to break or change the existing Uploader. 

The primary difference between them is that the UploadService is modal (because it uses a dialog) and the FileUploadService is not (because it uses the StickyPanel). We have no plans to deprecate the existing UploadService because it is now used in a product feature and there are no plans as yet to swap it out for the FileUploadService. It is also the case that people might want to use it over the FileUploadService.

Also, since then a customer has wanted to customize the UploadService and this resulted in the creation of https://issues.alfresco.com/jira/browse/AKU-851 which was fixed in 1.0.57.... so in actual fact it was *never* possible to customize the upload display in 1.0.51.

I haven't reviewed the details of the ticket that Martin has raised in relation to this, but it would definitely be useful if you (Axel) could try out against a more recent release (i.e. 1.0.57 onwards) to provide as much feedback and requirements on what improvements you'd like to see for customizing either or both upload services.

**Comment by AFaust** [comment]
The reason it didn't work for the customer in https://issues.alfresco.com/jira/browse/AKU-851 is that they were missing a crucial configuration property: the `assignTo : 'uploadDisplayWidget'` to have the display widget accessible for the BaseUploadService. So I wouldn't say it was _never_ possible to customize it - it only wasn't following clean separation of internal behaviour and external configuration.
For the project (which can't just update to every new Aikau release due to the rat's tail we've grown with regards to other fixes/patches and custom widgets), I'll just drop the mixin and rely on function name + argument contract.
Locally I'll do a quick&dirty update to a recent release to look a the new uploader stuff.

**Comment by draperd** [comment]
It's a shame that you're not able to update to newer versions because of the way you've patched earlier versions. Hopefully over time those patches can be removed and contributed back into the core project so that you can take advantage of the regular releases that Aikau provides. If not for the benefit of your current project, then maybe for future projects that you work on and also to help the Alfresco Community at large.

**Comment by MartinDoyleUK** [comment]
This code has been updated (see the referenced pull-request), and should be included in the next release of Aikau.

**Comment by draperd** [comment]
The fix for this is now available in the 1.0.60 release
